posts/2022/02/making-gpus-accessible-to-pydata-ecosystem-via-array-api.md

```diff
@@ -0,0 +1,481 @@
+  it fully generic.
+
+See for instance following piece of code:
+
+```python
+from scipy import sparse
+
+graph = sparse.coo_matrix(<input_arrays>)
+```
+
+Here the `sparse.coo_matrix` is a compiled code extension in the `scipy.sparse` module. We need the 
+dispatch to `cupyx.scipy.sparse.coo_matrix` for CuPy arrays and `scipy.sparse.coo_matrix` for NumPy
+arrays. This is not possible with the Array API's dispatching via `get_namespace`. For this demo we have
+added a workaround, until we have implemented the support for `uarray` based backed in SciPy. The
+workaround looks something like this for CuPy:
+
+```python
+xp, _ = get_namespace(<input_arrays>)
+
+if xp.__name__ == 'cupy.array_api':
+  from cupyx.scipy import sparse as cupy_sparse
+  coo_matrix =  cupy_sparse.coo_matrix
+else:
+  coo_matrix = scipy.sparse.coo_matrix
+
+graph = coo_matrix(<input_arrays>)
+```
+
+Note that, this workaround is only valid for NumPy/CuPy arrays.
+
+
+#### 2. Not implemented functions
+
+There are still a bunch of functions in the NumPy API which don't
+have a direct equivalent in the Array API, either due to the fact that those functions are not
+generic enough to be defined in the Array API or we haven't got around implementing those.
+
+Some examples of unimplemented functions that we encountered during this excercise:
+
+
+* `np.atleast_3d`
+* `np.strides`
+* `np.real`
+* `np.clip`
+
+#### 3. Inconsistencies with the NumPy API and Array API
+
+Although the Array API has managed to
+  get good coverage of the NumPy API, there are places where there are inconsistencies between
+  the two, it could be either with the function parameters or behaviour or the function or name
+  of the function like `concat` and `concatenate`.
+
+* Naming inconsistency between NumPy API and Array API
+
+```ipython
+
+In [1]: import numpy as np
+In [2]: import numpy.array_api as npx
+
+# Numpy API
+In [3]: np.concatenate
+Out[3]: <function numpy.concatenate>
+
+# Array API
+In [4]: npx.concat
+Out[4]: <function numpy.array_api._manipulation_functions.concat...>
+```
+
+* Indexing inconsistency between NumPy API and Array API
+
+```ipython
+In [1]: import numpy as np
+In [2]: import numpy.array_api as npx
+
+# Works with NumPy API
+In [3]: x = np.arange(10)
+In [4]: x[np.arange(2)]
+Out[4]: array([0, 1])
+
+
+# Doesn't Works with Array API
+In [5]: z = npx.arange(10)
+In [6]: z[npx.arange(2)]
+-----------------------------------------------------------------
+...
+IndexError: Non-zero dimensional integer array indices are not allowed in the array API namespace
+```
+
+## Results
+
+Here we will talk about about the performance of spectral clustering on GPU vs CPU
+
+We ran this on AMD as well as NVIDIA GPUs and here is a plot of the performance.
+The x-axis of the plot is the image dimension, i.e the dimension of the original
+image after rescaling it to a certain proportion, for this demo rescaled the image
+to 0.1, 0.2, 0.4, 0.6, 0.8 and 1x.
+
+
```

Comment:
Do you understand why smallest image sizes are slower on both AMD and NVIDIA GPUs? Is that real, or some kind of warm-up effect of the benchmarks?

Reply:
I don;t have a way to prove it, but as per my understanding on GPUs this is due to:
https://github.com/Quansight-Labs/quansight-labs-site/pull/302/files#diff-fa53d9a3136fe7fcd2fe2137246cd22993c1a5a69a9fb9af285e41663e3c3b3cR425-R426

Reply:
That explains why GPU is slower than CPU, but I don't think it can ever explain why a smaller image size takes more than a larger size on the same GPU.

Reply:
It's possible it's a real effect of course (something happens to work out better with cache size or an algorithm is switching behavior above a certain size). 

It's also possible that it's an artifact. For example, if you add an even smaller image size as the first data point, and then discard that new data point from the benchmarks, does the effect go away?

Reply:
Ah, yes that's true I misinterpreted your question. I don't have an idea about that.

Reply:
>  For example, if you add an even smaller image size as the first data point, and then discard that new data point from the benchmarks, does the effect go away?

That's a good point, let me try that.

Reply:
That looks much nicer and more sensible. There's still some noise, but overall this matches the expected behavior I'd say.